src/python/pants/backend/tools/semgrep/subsystem.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Iterable
+
+from pants.backend.python.goals import lockfile
+from pants.backend.python.goals.export import ExportPythonTool, ExportPythonToolSentinel
+from pants.backend.python.goals.lockfile import GeneratePythonLockfile
+from pants.backend.python.subsystems.python_tool_base import (
+    ExportToolOption,
+    LockfileRules,
+    PythonToolBase,
+)
+from pants.backend.python.target_types import ConsoleScript
+from pants.backend.python.util_rules.pex_requirements import GeneratePythonToolLockfileSentinel
+from pants.core.goals.generate_lockfiles import GenerateToolLockfileSentinel
+from pants.engine.rules import Rule, collect_rules, rule
+from pants.engine.target import Dependencies, FieldSet, SingleSourceField, Target
+from pants.engine.unions import UnionRule
+from pants.option.option_types import ArgsListOption, BoolOption, SkipOption
+from pants.util.docutil import git_url
+from pants.util.strutil import softwrap
+
+
+@dataclass(frozen=True)
+class SemgrepFieldSet(FieldSet):
+    required_fields = (SingleSourceField, Dependencies)
+    source: SingleSourceField
+    dependencies: Dependencies
+
+    @classmethod
+    def opt_out(cls, tgt: Target) -> bool:
+        # FIXME: global skip_semgrep field?
+        return False
+
+
+class Semgrep(PythonToolBase):
```

Comment:
I'd consider naming this `SemgrepTool` (or maybe `SemgrepSubsystem`).

We want to use suffix identification for target and field types etc, and I think that is a good pattern to apply here as well.